### ISSUE TITLE
Procedure not clear on how to remove/disable a language

### ISSUE THREAD

**Issue by Nafihahmd** [espeak-ng/espeak-ng#922]
Hi all,
I am trying to bring down the ROM/FLASH usage by the whole eSpeak system below 500 KB. For this I think it is necessary to remove unwanted languages. But the procedure to remove a language is no where laid out in the eSpeak documentation (or anywhere else in the internet),

So can somebody point me in the right direction to achieve this, please.
Nafih

**Comment by jaacoppi** [comment]
Duplicate of #918.

The language files are not included in the espeak-ng binary or library. They are stored separately in the espeak-ng-data directory. You can simply choose not to install the the unneeded _dict files.

You should also disable all configure options like sonic, speechplayer, klatt and mbrola. See ./configure --help.

Other than that I think your only option is to edit the source code. What I would do:
1. remove unnecessary language code in tr_languages.c
2. remove ssml support  in ssml.c if you don't need it
3. remove klatt and mbrola related code in klatt.c and mbrola.c and mbrowrap.c
5. remove all unneeded language specific checks in readclause.c, translate.c and numbers.c
6. Remove LoadConfig() and soundicon.c if you don't need it
7. remove unnecessary code related to language compiling. See compiledata.c and compiledict.c
Be careful with this one. Once this is done you can't produce new language files. If you've already done everything in parts 1-5, the original data files might not match your binary anymore.

**Comment by Nafihahmd** [comment]
Thank you for the quick reply. You gave a reply more detailed than I expected, thanks. I was editing the espeak-ng-data directory in the project folder (cloned folder). That is why it wasn't working. Reading your comment I remembered that there is another espeak-ng-data in `usr/share` directory and edited it. And yes it worked.

That means that to reduce size I have to think of other methods that you have mentioned. I had already disabled  all configure options. I will try the remaining procedures and report here before closing the issue.

Thanks

**Comment by jbowler** [comment]
That's not the whole picture, take a look at the code of src/libespeak-ng/tr_languages.c

$ wc src/libespeak-ng/tr_languages.c
1658  8976 72076 src/libespeak-ng/tr_languages.c

It's crude but if I look at the .o sizes I see this:

`   text    data     bss     dec     hex filename
  28496    3024   35552   67072   10600 src/libespeak-ng/la-compiledata.o
  16593    1256    8960   26809    68b9 src/libespeak-ng/la-compiledict.o
   1398       0       0    1398     576 src/libespeak-ng/la-compilembrola.o
  28305     280     272   28857    70b9 src/libespeak-ng/la-dictionary.o
   7760    2176       0    9936    26d0 src/libespeak-ng/la-encoding.o
   2139       0       0    2139     85b src/libespeak-ng/la-error.o
   2324       0       0    2324     914 src/libespeak-ng/la-espeak_api.o
   2476       0       4    2480     9b0 src/libespeak-ng/la-espeak_command.o
   3367       0     296    3663     e4f src/libespeak-ng/la-event.o
   4288       0     264    4552    11c8 src/libespeak-ng/la-fifo.o
   2719       0       0    2719     a9f src/libespeak-ng/la-ieee80.o
   6357    2528      40    8925    22dd src/libespeak-ng/la-intonation.o
   9251     736    4296   14283    37cb src/libespeak-ng/la-klatt.o
   6154       4     292    6450    1932 src/libespeak-ng/la-mbrowrap.o
    234       0       0     234      ea src/libespeak-ng/la-mnemonics.o
  21879     450      68   22397    577d src/libespeak-ng/la-numbers.o
   3982       5       0    3987     f93 src/libespeak-ng/la-phonemelist.o
   1946       0       0    1946     79a src/libespeak-ng/la-phoneme.o
  10974      12    3216   14202    377a src/libespeak-ng/la-readclause.o
   5212      12       4    5228    146c src/libespeak-ng/la-setlengths.o
   1494       0    1924    3418     d5a src/libespeak-ng/la-soundicon.o
   4030       0       0    4030     fbe src/libespeak-ng/la-spect.o
   7195      12     420    7627    1dcb src/libespeak-ng/la-speech.o
   2228       0       8    2236     8bc src/libespeak-ng/la-sPlayer.o
   8884    2000      44   10928    2ab0 src/libespeak-ng/la-ssml.o
   8335       0    9752   18087    46a7 src/libespeak-ng/la-synthdata.o
  15178       0   51112   66290   102f2 src/libespeak-ng/la-synthesize.o
   5611       0     120    5731    1663 src/libespeak-ng/la-synth_mbrola.o
  28815     500   10884   40199    9d07 src/libespeak-ng/la-translate.o
  18274    1064       0   19338    4b8a src/libespeak-ng/la-tr_languages.o
  16420     896    5332   22648    5878 src/libespeak-ng/la-voices.o
  10667    5056   22836   38559    969f src/libespeak-ng/la-wavegen.o
`

There are some big sore points in there.  The language related one is that even if you comment out all the languages you don't need in tr_languges.c the code it no longer requires is still in translate.c.

The bigger elephant is that the --compile support is still in libespeak.  compiledata.c, which is only concerned with compiling intonation and phoneme data, is the single biggest blob in there.

I did this many years ago (December 1996) with libpng - in the end I gave up and wrote it (libpng) from scratch.  It is possible to build software that is configurable in this way (I later did this with libpng) but espeak-ng has absolutely no configurability in this regard, so the only choice (seriously) is to fork it and eliminate everything you don't need.  You don't need the compilation stuff because, so far as I can see, if you can correctly build the dictionaries on your build platform you can just keep on doing that; so you build and do install-data using this repo but maintain a separate fork with your embedded version of the binaries, which have been hacked to remove all the compile support, all the language support you don't need etc.

Bear in mind you will have to produce flashes for each language set the product is sold on and they will differ significantly in size.  I suspect a useful contribution to espeak-ng would be a rewrite of translate.c to be entirely data driven.

**Comment by jbowler** [comment]
Hum... apart from, perhaps, Bosnian it seems to me that it would be possible to enable/disable languages in automake.  That's not going to help Windows users but it would allow an explicit language list (in the sense of a list of %_dict % values) to be specified and then maybe allow cpp defines to reduce the size of translate.c, though that is more difficult.

**Comment by jaacoppi** [comment]
You are absoulutely correct. Most users don't need the the language compilation code since it's only used once for installation. Based on the sizes of object files looks like all other changes would be almost meaningless compared to it. It could be configurable or even separated to a different binary.

**Comment by Nafihahmd** [comment]
Thank you guys for your valuable replies. I was working on setting up the cross compiling toolchain for a couple days hence the delay in my reply. And to be frank I haven't quite understood what @jbowler had suggested. But I will work on it from today onward. I have started a separate fork where my development will be carried out as you have suggested.

**Comment by jbowler** [comment]
So far as what @jaacoppi suggests it's pretty trivial to try - just remove src/libespeak-ng/compile*.c from the source files for the library/DLL...

I just tested it, the library builds with no issues.  I get (amd64 with -O2):

```
~/src/espeak-ng/master-build $ size !$/libespeak-ng.so.1.1.51
size src/.libs//libespeak-ng.so.1.1.51
   text    data     bss     dec     hex filename
 479817   31308  111464  622589   97ffd src/.libs/libespeak-ng.so.1.1.51
```
If I build the library with all the compile support:

```
~/src/espeak-ng/master-build $ size src/.libs//libespeak-ng.so.1.1.51
   text    data     bss     dec     hex filename
 533092   35708  155976  724776   b0f28 src/.libs//libespeak-ng.so.1.1.51
```

So that saves over 100k at run-time (text+data+bss).   ROM usage depends primarily on the text - data tends to be highly compressible - and that only amounts to 53275 bytes, still that's 10%.

An espeak-ng static build with the "compile" options commented out should just do this automatically, and commenting out the five compile options is easy too...

**Comment by Nafihahmd** [comment]
Hi @jbowler , as you said I removed all compile*.c tried recompiling. But it is giving me errors. Let me elaborate the steps I followed.

1. I moved compile*.c files out of the main folder
  ```
  mv src/libespeak-ng/compile*.c ../
  ```
2. I recompiled the code using *make -B*
3. Now in the middle of compiling I am getting the following error.
```
  CC       src/ucd-tools/src/src_libespeak_ng_la-tostring.lo
make[1]: *** No rule to make target 'src/libespeak-ng/compiledata.c', needed by 'src/libespeak-ng/src_libespeak_ng_la-compiledata.lo'.  Stop.
make[1]: Leaving directory '/home/melayil/Github/espeak-ng'
Makefile:749: recipe for target 'all' failed
make: *** [all] Error 2
```

*N.B*: I get similar error if I remove ssml.c file also (make[1]: *** No rule to make target 'src/libespeak-ng/ssml.c,...')

*N.B 2*: I think @jaacoppi was asking to removing only parts of .c file in his previous post. If that is the case, how should I remove functions inside the .c files without effecting other code that will be calling the former function. Should I leave a dummy function with no code inside?

**Comment by jaacoppi** [comment]
The build system (Gnu Make) expects those files to exist. You'll get an error if you remove them.

One way to do it is to delete everything, just leave an empty file. Then recompile and wait for the compiler to give error messages about missing functions. Then create dummy functions for those parts.

Another way is to look through the code and figure out where each function is called and why. Then delete everything you don't need.

**Comment by Nafihahmd** [comment]
Your method was very quick buddy. Removed SSML functionality. I just had to add 2 dummy functions. I will write the new ssml.c below.

```


#include "config.h"

#include <ctype.h>
#include <errno.h>
#include <locale.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <wchar.h>
#include <wctype.h>


#include <espeak-ng/espeak_ng.h>
#include <espeak-ng/speak_lib.h>
#include <espeak-ng/encoding.h>
#include <ucd/ucd.h>

#include "ssml.h"
#include "dictionary.h"           // for strncpy0
#include "mnemonics.h"               // for LookupMnemName, MNEM_TAB, 
#include "readclause.h"           // for PARAM_STACK, param_stack, AddNameData
#include "soundicon.h"               // for LoadSoundFile2
#include "synthesize.h"           // for SPEED_FACTORS, speed
#include "translate.h"            // for CTRL_EMBEDDED, IsDigit09, utf8_out
#include "voice.h"                // for SelectVoice, SelectVoiceByName

static MNEM_TAB ssmltags[] = {
	{ "speak",     SSML_SPEAK },
	{ "voice",     SSML_VOICE },
	{ "prosody",   SSML_PROSODY },
	{ "say-as",    SSML_SAYAS },
	{ "mark",      SSML_MARK },
	{ "s",         SSML_SENTENCE },
	{ "p",         SSML_PARAGRAPH },
	{ "phoneme",   SSML_PHONEME },
	{ "sub",       SSML_SUB },
	{ "tts:style", SSML_STYLE },
	{ "audio",     SSML_AUDIO },
	{ "emphasis",  SSML_EMPHASIS },
	{ "break",     SSML_BREAK },
	{ "metadata",  SSML_IGNORE_TEXT },

	{ "br",     HTML_BREAK },
	{ "li",     HTML_BREAK },
	{ "dd",     HTML_BREAK },
	{ "img",    HTML_BREAK },
	{ "td",     HTML_BREAK },
	{ "h1",     SSML_PARAGRAPH },
	{ "h2",     SSML_PARAGRAPH },
	{ "h3",     SSML_PARAGRAPH },
	{ "h4",     SSML_PARAGRAPH },
	{ "hr",     SSML_PARAGRAPH },
	{ "script", SSML_IGNORE_TEXT },
	{ "style",  SSML_IGNORE_TEXT },
	{ "font",   HTML_NOSPACE },
	{ "b",      HTML_NOSPACE },
	{ "i",      HTML_NOSPACE },
	{ "strong", HTML_NOSPACE },
	{ "em",     HTML_NOSPACE },
	{ "code",   HTML_NOSPACE },

	{ NULL, 0 }
};

/********* Dummy Functions *********/

int ProcessSsmlTag(wchar_t *xml_buf, char *outbuf, int *outix, int n_outbuf, const char *xmlbase, bool *audio_text, char *current_voice_id, espeak_VOICE *base_voice, char *base_voice_variant_name, bool *ignore_text, bool *clear_skipping_text, int *sayas_mode, int *sayas_start, SSML_STACK *ssml_stack, int *n_ssml_stack, int *n_param_stack, int *speech_parameters)
{
	return 0;
}

int ParseSsmlReference(char *ref, int *c1, int *c2) {

	return -1; //Failure to Parse

}
```

And now to the real business.
With SSML
|file                      | size (bytes) |
|libespeak-ng.a   |2,584,008     |
|libespeak-ng.so |1,557,248      |

Without SSML
|file                      | size (bytes) |
|libespeak-ng.a   |2,516,848     |
|libespeak-ng.so |1,514,088      |

**Comment by jbowler** [comment]
You don't remove the files, rather you modify Makefile.am not to include them in the build; i.e. delete the three lines:

```
        src/libespeak-ng/compiledata.c \
        src/libespeak-ng/compiledict.c \
        src/libespeak-ng/compilembrola.c \
```
from the definition of src_libespeak_ng_la_SOURCES and whichever of the phsource/ph_* lines and espeak-ng-data/*_dict lines that you don't need to build.  Although, since espeak_ng won't be able to do the compile operations you probably don't even want to bother with the latter; just comment out the 'all-local' target at the top of Makefile.am

For tr_translate.c you have to comment out the languages.  BTW this is for an experiment or a one-time unsupportable hack for a particular system.  Actually doing it in the project would require some way of choosing the languages at 'configure' time, some support for the same in the Visual Studio solution and extra configure options such as --enable-ssml plus a somewhat tricky change to allow espeak-ng to be built twice, once with compile support once without, or maybe --enable-compile.

**Comment by jbowler** [comment]
Yes, that's expected.  Now you need to track down the internal references in libespeak-ng; the print_dictionary_flags sounds innocent, but the fact that "DecodeRule" is trying to compile a dictionary seems really weird to me.  Of course, I'm saying this on the basis of the function names...

Don't you have your own app?  Ultimately if you are reducing code usage in a library that's what matters; the library compiles fine but it clearly has missing references that should be resolved internally.  With your own app you can do a static link and remove whole functions from the library.

**Comment by Nafihahmd** [comment]
@jaacoppi I have removed unwanted languages from tr_languages.c and the result is :


All Languages*

|file                      | size (bytes) | size after stripping** |
---------------------  |  ---------------- | -----------------------
|libespeak-ng.a   |2,584,008     |454,842  |
|libespeak-ng.so |1,557,248      | 498,486 |


English Only

|file                      | size (bytes) | size after stripping |
---------------------  |  ---------------- | ---------------------
|libespeak-ng.a   |2,420,760     | 440,866 |
|libespeak-ng.so |1,448,576      | 481,016 |

English + South Indian Languages

|file                      | size (bytes) | size after stripping |
---------------------  |  ---------------- | --------------------
|libespeak-ng.a   |2,427,696     | 441,954 |
|libespeak-ng.so |1,456,096      | 485,240 |


\* All languages describe size after the removal of SSML feature  
\*\* Strip means remove debug information from a library. It is done using **Strip** command.  
eg., `strip libespeak-ng.a`

**Comment by Nafihahmd** [comment]
I have removed unwanted languages all language and the result is :


**Before**:

|file                      | size (bytes) | size after stripping** |
---------------------  |  ---------------- | -----------------------
|libespeak-ng.a   |2,584,008     |454,842  |
|libespeak-ng.so |1,557,248      | 498,486 |

**After**:

|file                      | size (bytes) | size after stripping** |
---------------------  |  ---------------- | -----------------------
|libespeak-ng.a   |2,107,022     |392,106  |
|libespeak-ng.so |1,264,176      | 427,736 |

After removing `compiledict.c` you will not be able to compile espeak-ng from scratch. For compiling I manually placed compiled `espeak-ng-data` folder (taken from another build without compile data removed) inside the project folder.  

I have a doubt as well. It's a novice one. What feature is removed as I remove compiledata.c ?

**Comment by jbowler** [comment]
Since you had to dummy out functions outside the three compile*.c files presumably something is now missing.  Put an 'fprintf(stderr)' into the function dummies and run the full set of tests excluding ones that are compile specific (there may not be any such tests).  Do this on a build with all the languages; just the compilation changes.  It's probably easiest to start with "make -k check" and see what happens.  Check the output for the dummy 'fprintf'.

The sequence after putting the fprintf's in is something like:

cd test directory; make all # this will fail at the first compile
cd other-build; make -B
cp -rp other-build/espeak-ng-data test-directory/
cd test-directory; make -k check

I'm looking at a way of making this easier; to allow no-compile builds to work with "make all".  It's a requirement for making cross-builds work, but that might take a while even if my pull requests are accepted...

You could also publish the changes to a git archive somewhere or, since they are in a preliminary state, "gif diff origin/master" (or the equivalent, depending on how you made the clone) and add the result as an attachment to this request.  Then people who know the code well can probably instantly say if there will be a problem.

**Comment by jbowler** [comment]
"make check" currently fails because of this:

make: *** [Makefile:2842: tests/translate.check] Error 1

I verified that the problem is  in the current "master" branch

**Comment by jaacoppi** [comment]
The #824 test in tests/translate.test is known to produce different results depending on the compiler used. Same for tests/klatt.test. There's no fix yet.

**Comment by jbowler** [comment]
I git bisected the problem to dca948f5f1decb51dd09bc6c3c142dfbcef375d6, that change is the "simple" movement of a single file in the tree; the prior build passes "make check" and the build with the file move fails.  This is with an identical compiler (obviously).  Reverting the change higher up the tree doesn't fix the problem so it isn't simply the change (corrrection) in the Belta lang file location.

I'm still working on narrowing down where the problem comes from.  Conceivably, given the language additions require code changes, it might be a language addition.  IRC moving the "lang" file to the right place will have turned the language "on".

**Comment by jbowler** [comment]
Hum...  Ok, it was the following commit - "add test for #824" - the problem is that this causes "make check" to fail which makes it impossible to create pull requests for any significant source code change, or, indeed, check beyond 'translate.test'...

**Comment by jaacoppi** [comment]
Indeed. The question is why does the test fail on some compilers and pass on others. Removing the test is not an option since it's needed to prevent #824.

**Comment by jbowler** [comment]
Yes but it prevents all the other tests running; for example by marking the #824 test "broken" I get to see a whole run of errors in tests/ssml.test (which I haven't tracked down yet).  I could supply a patch to add MESSAGE "soft" to test_phon and if espeak-ng does not crash the test succeeds even if the output is 'wrong'.

Anyway, the error is really weird - somehow the incorrect version is adding a # character in certain very particular places:

ka v'os;  --> ka# v'os;
ka vOs; --> ka# vOs;
ka t@-r; --> ka# t@-r;
ka S,E#s; --> ka# S,E#s;t;
'ista d; --> 'ista# d;
E2sta d; --> E2sta# d;
ka n'ojldv'A_! tR; --> ka# n'ojldv'A_! tR;
'ista d; --> 'ista# d;

I.e. somehow the "broken compiler" case is outputing "a " as "a# " - there are not instances of "a# " in the correct output.   I'll file a bug with a patch to sed the output of this one test.

**Comment by graduatemazada** [comment]
i would like to know how to delete the code in tr_language.c

**Comment by jaacoppi** [comment]
@graduatemazada just delete the parts you don't need. If they affect any tests, delete those tests.

**Comment by thewh1teagle** [comment]
> The language files are not included in the espeak-ng binary or library. They are stored separately in the espeak-ng-data directory. You can simply choose not to install the the unneeded _dict files.

But how to disable them during cmake compile? for instance I want to exclude everything in dictsource/extra
